### ISSUE TITLE
"Test approach" vs. "test practice"

### ISSUE THREAD

**Issue by samm82** [samm82/TestingTesting#21]
# Main Point
I think I'll change the "Approach" value to "Practice", "Design" to "Technique", and the "Field of Testing" column header to "Approach Category".

## Reasoning

Following up on discussion in #19, I realized that the definition of "test approach" includes the following note: "Typical choices made as test approaches are _test level_, _test type_, _test technique_, _test practice_ and the form of _static testing_ to be used" (IEEE, 2022, p. 10). "Test technique" is a synonym for "test design technique" (p. 11), which I called "design" in the "Field of Test" column of my glossary. So it seems like the general phrase for all the entries in my glossary are "approaches", and the "generic" ones are actually "test practices": "conceptual framework[s] that can be applied to the organizational test process, the test management process, and/or the dynamic test process to facilitate testing" (p. 14). This is supported by the following uses:

- "the test strategy specifies that the test practices scripted testing, exploratory testing and test automation are use[d]" (p. 20)
- Figure 2 on p. 22 supports this breakdown of "test approach"
- "experience-based test practices are primarily unscripted" and "include (but are not limited to) exploratory testing, tours, attacks, and checklist-based testing" (p. 34)
- "Mathematical-based test practices can be used...." (p. 36)

I'm not sure where we actually landed on the term "test practice", since we may have glossed over it, but I just wanted to keep you in the loop/get your feedback.

# What about Static Testing?

This brings up the question I asked in the meeting of "why did they group static testing separately in Figure 2 on p. 22?" I think the answer is in their understanding of test approach. However, I'm not convinced that this distinction is quite correct; you could argue that a lot of the design techniques are "static", since you don't need to run the code to analyze its equivalence partitions, for example. Therefore, I think the four "Types of Approach" are "Level", "Type", "Design Technique" (likely shortened to "Technique" in the glossary), and "Practice", with "Practice" being a sort of catch-all (including most, if not all, of the "static" approaches).

**Comment by smiths** [comment]
I'm sure what you have said above @samm82 is internally consistent, but I'm finding it hard to distinguish all of the words.  Practice versus approach versus technique etc.  Can you summarize your terminology with a definition for each of your phrases?  I'd really like an example of each too.  I find I understand things better with an example.  :smile:

**Comment by samm82** [comment]
# Definitions from IEEE

- **Approach:** A "high-level test implementation choice, typically made as part of the test strategy design activity" that includes "test level, test type, test technique, test practice and the form of static testing to be used" (p. 10) (e.g., any of the examples given below: equivalence partitioning, unit testing, scripted testing, functional testing)
- **(Design) Technique:** "procedure used to create or select a test model, identify test coverage items, and derive corresponding test cases" (e.g., equivalence partitioning, boundary value analysis, branch testing) (p. 11)
- **Level:** A stage of testing "typically associated with the achievement of particular objectives and used to treat particular risks" (e.g., unit/component testing, integration testing, system testing) (p. 12)
- **Practice:** A "conceptual framework that can be applied to …[a] test process to facilitate testing" (p. 14) (e.g., scripted testing, exploratory testing, automated testing) (p. 20)
- **Type:** "Testing that is focused on specific quality characteristics" (e.g., functional testing, usability testing, and performance testing) (p. 15)

_Side note: where's a good place for these? I decided to list them in the Methodology section of my notes for now; should I add them at the bottom of my spreadsheet as well?_

It also gives, as an example of a test approach "the use of model-based testing for the functional system testing" (p. 10). Note that there are three approaches--model-based testing, functional testing, and system testing--each of a different _kind_ of approach--practice, type, and level, respectively. This is consistent with how "model-based testing" is described as a "test practice" as a subset of "test approach" in Fig. 2 on p. 22.

# Definitions in My Own Words

- **Approach:** One of the below
- **(Design) Technique:** What/where the test case is derived from
- **Level:** What part of the artifact(s) is/are tested **OR** When in the process the test occurs
- **Practice:** How the test is structured/executed
- **Type:** Which quality is being tested (for?)

I'm a little iffy on my first definition of "level", since "part" is ambiguous; are branches "parts"? does this mean branch testing is a test level? (Spoiler: it's not.)

_Note: I implicitly exclude "the form of static testing to be used" as a subset of test approach because I don't think that makes sense as a distinction at this level; as far as I can tell, any "form of static testing" can be described by another kind of test approach._

**Comment by samm82** [comment]
Depending on the scope of which test activities are considered "in-scope" vs. "out-of-scope" as supporting activities, it may be beneficial to also introduce a category of "metric" to described things like coverage and error seeding.

**Comment by samm82** [comment]
Interesting to note that Barbosa et al. (2006, p. 2) have a different terminology: testing process, phase, artifact, step, procedure, and resource. As mentioned before, "phase" is synonymous to "level", and "artifact" and "resource" are not part of this specific discussion (although will come up in later discussion). This leaves the following terms from (p. 3):

- **Process:** "based on a development technology and … paradigm, as well as on a testing procedure"
- **Step:** A subset of a testing process, such as planning, construction, execution, or analysis
- **Procedure:** "categorized in[to] testing methods, testing guidances and testing techniques"

From these, "process" seems to map to IEEE's "practice" and "procedure" seems to map to IEEE's "approach", the overarching category. "Methods" and "guidances" are not defined, but my intuition tells me that "method" may be a synonym of "process", and "guidance" may be related to/part of either a new category of "[metric](https://github.com/samm82/TestGen-Thesis/issues/21#issuecomment-1899172350)" or the category described here as "technique", which is described as a way to "perform the tests in a systematic way and on a sound theoretical basis (p. 3)"; this seems to map to IEEE's "technique" (yay! consistency!)

**TL;DR:** I personally think that IEEE's distinctions are more clear, but this definitely highlights the possibility/need for setting up a consistent nomenclature for categorizations of testing.

**Comment by samm82** [comment]
I've added tables to my notes doc on what terms there are and where they come from; hopefully this makes things clearer!

![IEEE terms](https://github.com/samm82/TestGen-Thesis/assets/35857611/5c44fe9e-a5fb-4d82-b0ab-ff4e1dc02634)

![other terms](https://github.com/samm82/TestGen-Thesis/assets/35857611/a7eb8b9e-81a3-45bf-a7ff-26e235cd6ac9)

**Comment by smiths** [comment]
Lots of good content here @samm82.  Nice to see you providing traceability to the source of the definitions.  The distinctions/definitions you summarize and make seem good to me.  Of course, I reserve the right to change my mind.  The real test will be when we use these terms how much do they help us build up a taxonomy that resonates.

There are some little points that seem odd to me.  For instance, the definition of Type of testing talks about focusing on a specific quality, but then the example includes functional testing.  Functional testing doesn't test qualities, but functionality.  That is, it tests what the software does, not the quality of how well it does it.  The other examples usability and performance are qualities, but functionality is not a quality.

You asked above where a good place is to put your definitions.  The methodology section you mentioned is a good place for now, although we might eventually move them to a glossary section.  I don't think you should repeat the definitions in your spreadsheet.  We don't want to deal with the maintenance headache of putting information in two different places.

**Comment by JacquesCarette** [comment]
I must admit that I find all of these word/definition pairs unintuitive and not necessarily helpful (even though the quality of the underlying data gathering and reporting is excellent). But they do seem to serve a real purpose, so I'll see if I can get used to them.

**Comment by smiths** [comment]
I agree @JacquesCarette.  That is why I don't think we should put too much time into debating the words and definitions.  We could go that route, but it will take time and at some point, it won't be fun anymore.  :smile:  I think we adopt the terminology that  @samm82 has identified and see how well it serves our purposes in the big summary table.  If it isn't adequate in some way, we'll find out while we are using it.  If there is more terminology than we need, we should be able to notice that too.

**Comment by samm82** [comment]
That's actually a good point. I figured the first "approach vs. design" distinction would be helpful, but as this category became more fleshed out, I was wondering how much of this discussion would fit more in the analysis stage. I'll close this for now, continuing to record anything of note I find, and we can revisit this discussion later if things don't sort themselves out.
